### PR TITLE
fix(history): skip nested repositories unless tracked directly

### PR DESCRIPTION
**Pull request by jdx**
https://entire.io/gh/jdx/mise/trails/62

Repositories found inside tracked directories are now skipped and reported instead of saved as Git commit pointers. A pointer did not contain the repository's files and could make adoption or pull fail when its commit was unavailable.

To save a vendored plugin's working files, track its root explicitly:

```sh
mise dot track ~/.hammerspoon
mise dot track ~/.hammerspoon/Spoons/SkyRocket.spoon
```

The first command reports the nested repository as skipped. The second captures its working files under its own tracking policies, always excluding `.git`. Save, status, paths, and history output explain skipped repositories, including JSON output.

Existing pointer-only history changes are skipped during pull/adoption without creating, removing, or updating a local checkout. New captures do not create those pointers. Skipped repositories are also recorded as omissions in shared checkpoint metadata, so rebuilding the local cache retains the information rollback needs to preserve their files.

Validation: existing CI passed on the reviewed head. Tests cover omission reporting, explicit repository-root tracking, legacy pointers, multi-machine adoption, and rollback after removing a nested `.git`.

Documentation validation: scoped formatting/lint checks, TOML example parsing, and a full site build at the stack tip. The five focused dotfiles end-to-end tests passed locally.

*AI-assisted — Tool: Codex; model: OpenAI/unavailable; version: unavailable.*

---

> [!NOTE]
> **Medium Risk**
> Changes dotfile capture, sync reconciliation, and rollback semantics for nested `.git` trees and legacy gitlinks; mistakes could affect multi-machine setups and data preservation during pull/rollback.
> 
> **Overview**
> **Nested Git directories under a tracked tree are no longer saved as commit pointers.** mise skips them entirely, surfaces them as `nested:` in save/status/paths/history (including JSON), and tells users to **`mise dot track <repo-root>`** to capture working files without `.git`.
> 
> **Sync and adoption** treat legacy `160000` pointers as non-content: reconcile skips apply/remove for pointer-vs-pointer cases, records **not shared** reasons in sync status, blocks taking remote pointer versions on pull, and still applies ordinary files while leaving live nested checkouts untouched.
> 
> **Rollback and checkpoints** persist skipped nested paths in coverage (and mirror them into shared commit omissions for older clients), so rollbacks do not delete paths the checkpoint never held—even after `.git` is removed. The watcher stops autosaving paths inside an untracked nested repo.
> 
> Docs and e2e tests cover the new behavior, legacy pointers, multi-machine bootstrap, and rollback edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1ac20f17cf16511f9e89b4b1f13a3939f70178d5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/system/history/checkpoint.rs [hunk 1]
```diff
@@ -578,6 +580,27 @@ impl Store {
...
+        tree: &str,
+        walk: &super::tracked::Walk,
+    ) -> Result<bool> {
+        if walk.nested.is_empty() && previous.tree.coverage.nested.is_empty() {
+            return Ok(false);
+        }
+        let Some(repo) = &self.repo else {
+            return Ok(false);
+        };
+        Ok(repo.with_pointers_of(snapshot, tree)? == tree)
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/checkpoint.rs]: Held: `compose` removes everything under an overlaid path, so the previous pointer replaced the new subtree in the comparison. Fixed in 8a6beb152: `with_pointers_of` substitutes a pointer only against another pointer or nothing; content on either side is left alone and compares as a real change. Unit test on the tree helper; the e2e removes the nested `.git`, saves, and pulls the files on another machine.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/shadow.rs [hunk 2]
```diff
@@ -972,6 +981,33 @@ impl HistoryRepo {
...
+            }
+        }
+        let overlays: Vec<Overlay> = pointers
+            .into_iter()
+            .map(|(path, object)| Overlay { path, object })
+            .collect();
+        if overlays.is_empty() {
+            return Ok(tree.to_string());
+        }
+        self.compose(tree, &overlays)
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/shadow.rs]: Held for the comparison and for `is_pointer_conflict` (the reconcile skip keys on the flat path map, where a directory has no object at the pointer path, so it was already a per-path skip only). Fixed in 8a6beb152: pointer substitution and the conflict filter apply only when both sides are a pointer or absent; content against a pointer is a real difference and goes through the ordinary merge/hold path.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/checkpoint.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/checkpoint.rs]: Held; fixed in 8a6beb152 as described on the sibling threads (no substitution where the other tree has content), with a unit test on `with_pointers_of` and an e2e step that removes `.git` and saves immediately.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/reconcile.rs [hunk 3]
```diff
@@ -196,6 +227,19 @@ pub(crate) fn reconcile(
...
+        // it would delete a repository. It is never applied, never removed,
+        // and never a conflict; the pointer is published with the tree.
+        if is_gitlink(s) || is_gitlink(t) {
+            plan.skipped = Some(NESTED_NOT_SHARED.into());
+            plan.next.acknowledged = t_version.clone();
+            plan.next.reconciled = t_version.clone();
+            plan.next.applied = t_version;
+            plans.push(plan);
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/reconcile.rs:235]: Held, and worse than described: the merge is clean (the pointer side is unchanged against base), so the files would have been written into the still-nested checkout, and that machine's next capture would then re-record its pointer and publish it over the files, deleting them upstream. Fixed in 3d9d815b4: repository application and publication compare the local tree against the remote tree and pause at a path that is a pointer on one side and content on the other, naming the way out (remove the local `.git`, or restore the repository on the other machine); a skipped pointer plan no longer counts as an accepted conflict resolution. Unit test on the tree check; the e2e has the still-nested machine record its pointer again and asserts both `sync` and `pull` pause, its checkout stays clean, and the files stay upstream.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/publish.rs [hunk 4]
```diff
@@ -30,10 +30,27 @@ pub(crate) fn build(
...
     }
     let tree = repo.output_tree_of(local)?;
+    // a repository that became ordinary files on another machine while it
+    // is still checked out here (or the reverse) must not be published
+    // over the other side, whichever of the two is ahead
+    if let Some(remote) = &heads.remote
+        && let Some(path) = repo.pointer_content_mismatch(&tree, &repo.output_tree_of(remote)?)?
+    {
+        bail!(
+            "sync paused: {path} is a nested repository on one machine and ordinary files on another; remove its .git here to receive the files, or restore it on the other machine"
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/publish.rs]: Held: the check ran over the whole trees before `eligible`. Fixed in fab4ab2ff: the pause is decided in `run.rs` (`live_pointer_mismatch`), only for paths this machine selects, and only when the local path still has a live `.git`; `publish::build` no longer checks on its own.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/shadow.rs [hunk 5]
```diff
@@ -972,6 +981,82 @@ impl HistoryRepo {
...
+                }
+                if let Some((mode, _)) = self.object_at(content, &entry.path)?
+                    && mode != "160000"
+                {
+                    return Ok(Some(entry.path));
+                }
+            }
+        }
+        Ok(None)
+    }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/shadow.rs]: Held in part: a machine that converted the repository itself was indeed paused whenever upstream still carried its old pointer (the e2e happened to run with the pointer absent upstream); a machine without the checkout was not, since its tree holds no pointer. Fixed in fab4ab2ff: the check is direction-specific (local pointer against remote content only), filtered by `eligible`, and requires a live `.git` at the local path, so the converting machine publishes, a machine that removed its checkout receives the files, and only a live checkout pauses. The apply's own consistency check also compares modulo pointers now, since the protective capture records the checkout's absence. The e2e covers the paused checkout and the files arriving once it is removed.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/run.rs [hunk 6]
```diff
@@ -962,6 +994,43 @@ fn apply_resolutions(
...
+/// no longer has it on disk, is not held up.
+fn live_pointer_mismatch(
+    repo: &crate::system::history::shadow::HistoryRepo,
+    tracked: &TrackedSet,
+    local: Option<&str>,
+    remote: Option<&str>,
+) -> Result<Option<String>> {
+    let (Some(local), Some(remote)) = (local, remote) else {
+        return Ok(None);
+    };
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/run.rs]: Held: with no local history the check returned before looking at the disk, and the existing per-path conflict was the only thing standing between the pull and the checkout. Fixed in ec65a942d: the pause is decided in `prepare` for every plan, and without local history it checks the remote content paths against the filesystem for a live `.git` between the entry root and the file; the e2e adopts on a fresh machine that already has the repository checked out and asserts the adoption pauses with the checkout untouched.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/reconcile.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/reconcile.rs:235]: Held: a type-change conflict at the pointer path was unresolvable once another machine had shipped a different pointer. Fixed in ec65a942d: in publication, a conflict whose remote side is a pointer is resolved with this machine's content (a live checkout here was already paused before publication runs), so the converting machine publishes its files; the e2e has another machine publish its pointer at a different commit first, then converts and syncs.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: e2e/cli/test_dotfiles_nested_repository [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+wrapper "$fourth" "$PWD/fourth"
+d="$PWD/fourth"
+mkdir -p "$fourth/.hammerspoon/Spoons/Sky.spoon"
+echo 'return { fourth = true }' >"$fourth/.hammerspoon/Spoons/Sky.spoon/init.lua"
+git -C "$fourth/.hammerspoon/Spoons/Sky.spoon" init -q -b main
+git -C "$fourth/.hammerspoon/Spoons/Sky.spoon" -c user.name=t -c user.email=t@example.com add -A
+git -C "$fourth/.hammerspoon/Spoons/Sky.spoon" -c user.name=t -c user.email=t@example.com commit -q -m fourth
+assert_fail "$d bootstrap --adopt file://$origin --yes --only dotfiles 2>&1" 'Sky.spoon is a nested repository here and ordinary files on another machine'
+assert "cat $fourth/.hammerspoon/Spoons/Sky.spoon/init.lua" 'return { fourth = true }'
+assert "git -C $fourth/.hammerspoon/Spoons/Sky.spoon status --porcelain | wc -l | tr -d ' '" 0
```
[the remark this answers is not in the document]
Comment (jdx) [reply e2e/cli/test_dotfiles_nested_repository]: This PR was reduced to reporting and skipping the pointer; the checkout-status assertions this refers to (and the sync pause they exercise) moved to the follow-up https://github.com/jdx/mise/pull/13422, where the e2e keeps the fixture's own `.git` (`git -C ... init` plus a commit) and where the stricter identity check belongs.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/run.rs [hunk 8]
```diff
@@ -962,6 +1009,68 @@ fn apply_resolutions(
...
+    repo: &crate::system::history::shadow::HistoryRepo,
+    tracked: &TrackedSet,
+    local: Option<&str>,
+    remote: Option<&str>,
+) -> Result<Option<String>> {
+    let Some(remote) = remote else {
+        return Ok(None);
+    };
+    let roots = Roots::current();
+    let Some(local) = local else {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/run.rs]: `live_pointer_mismatch` no longer exists on this branch: the sync machinery moved to the follow-up https://github.com/jdx/mise/pull/13422. Noted there as a known limitation for the maintainer (a repository initialised after the last save is not a pointer in either tree, so only the per-file conflict path stands between an incoming edit and the checkout).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/reconcile.rs [hunk 9]
```diff
@@ -196,6 +207,19 @@ pub(crate) fn reconcile(
...
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/reconcile.rs:235]: Holds, and it is exactly the class the follow-up https://github.com/jdx/mise/pull/13422 handles (`live_pointer_mismatch` pauses application and publication while a live `.git` stands at a path another machine published files under). This PR is deliberately reduced to reporting and skipping the pointer; on this branch a differing file inside the checkout is still held by the per-file conflict path, an identical one is adopted as-is. Recorded in the body as the trade-off and in https://github.com/jdx/mise/pull/13422 as the fix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/run.rs [hunk 10]
```diff
@@ -607,9 +617,31 @@ fn record_pending(
...
+            Some(SkippedPath {
+                branch_path: plan.branch_path.clone(),
+                reason,
+            })
+        })
+        .collect();
     status.pending_applications = plans
         .iter()
-        .filter(|plan| plan.conflict.is_none())
+        .filter(|plan| plan.conflict.is_none() && !pointer_application(plan))
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/run.rs]: Held: the plan kept `apply` set, and the sync loop waits on any application before publishing. Fixed in 27e07992b: `reconcile` turns an application that would write a pointer into a skipped, acknowledged no-op (`next` records the pointer as applied), so it is neither pending nor a block on publication; unit test covers acknowledged content that became a repository upstream.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/run.rs]: Held; fixed in 27e07992b as on the sibling thread: the pointer application is cleared and acknowledged in `reconcile` itself, so publication proceeds.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/run.rs [hunk 11]
```diff
@@ -607,6 +617,15 @@ fn record_pending(
...
+        .filter_map(|plan| {
+            Some(SkippedPath {
+                branch_path: plan.branch_path.clone(),
+                reason: plan.skipped.clone()?,
+            })
+        })
+        .collect();
     status.pending_applications = plans
         .iter()
         .filter(|plan| plan.conflict.is_none())
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/run.rs:631]: Held: a resolution could put the remote pointer back on the plan after reconcile's pass. Fixed in c9b58a9d5: the skip is a function (`skip_pointer_applications`) run after every `apply_resolutions`, so taking the remote side of a content-versus-pointer conflict records the pointer as acknowledged instead of writing it; unit test covers a decided plan.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/run.rs [hunk 12]
```diff
@@ -766,6 +786,7 @@ fn prepare(
     };
     let mut plans = reconcile_set(tracked)?;
     apply_resolutions(repo, status, shared, upstream, &mut plans)?;
+    reconcile::skip_pointer_applications(&mut plans);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/run.rs]: Held: taking the remote side of a pointer conflict recorded the pointer as acknowledged and then published the local files, silently contradicting the choice. Under this PR's rule (sync never writes a pointer) the honest answer is to refuse the decision: fixed in 85e874fa2, `--take-remote` on a path whose repository version is a nested repository pointer fails at decision time with the way out named (`--keep-local`, or replace the files with a clone of that repository yourself), so no decision is recorded that cannot take effect.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/store.rs [hunk 13]
```diff
@@ -388,6 +388,13 @@ impl Checkpoint {
                 .iter()
                 .filter_map(|item| portable(&item.path))
                 .collect(),
+            nested: self
+                .tree
+                .coverage
+                .nested
+                .iter()
+                .filter_map(|item| portable(&item.path))
+                .collect(),
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/store.rs]: Both correct, and fixed in 79f34c22b by not touching the shared format at all. `CommitRecord` uses `deny_unknown_fields`, so any new trailer field stops released clients reading this history after a pull — which is a much worse outcome than the reporting gap it was closing.

The list of skipped repositories now stays in the machine's own checkpoint record: the record is still read back from Git so every machine reads the same shared one, and this one field is carried across that round trip locally. That is where it is needed — a rollback consults it on the machine doing the rollback — and other machines find their own nested repositories the same way, by looking at their own filesystem. The trailer is unchanged, so an older mise reads this history exactly as before.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/store.rs [hunk 14]
```diff
@@ -441,6 +448,11 @@ pub(crate) struct CommitRecord {
     pub omitted: Vec<String>,
     /// Portable directory prefixes whose inventory was incomplete.
     pub incomplete: Vec<String>,
+    /// Portable paths of repositories this commit skipped. Nothing is
+    /// written to the tree for them, so the trailer is the only record
+    /// that they were there and were left alone.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub nested: Vec<String>,
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/store.rs]: Both correct, and fixed in 79f34c22b by not touching the shared format at all. `CommitRecord` uses `deny_unknown_fields`, so any new trailer field stops released clients reading this history after a pull — which is a much worse outcome than the reporting gap it was closing.

The list of skipped repositories now stays in the machine's own checkpoint record: the record is still read back from Git so every machine reads the same shared one, and this one field is carried across that round trip locally. That is where it is needed — a rollback consults it on the machine doing the rollback — and other machines find their own nested repositories the same way, by looking at their own filesystem. The trailer is unchanged, so an older mise reads this history exactly as before.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/replay.rs [hunk 15]
```diff
@@ -1290,6 +1301,16 @@ enum PathState {
...
+/// The repository a path lies in, when that repository is itself inside a
+/// tracked directory rather than tracked in its own right.
+fn nested_repository_at_or_above(tracked: &TrackedSet, path: &Path) -> Option<PathBuf> {
+    let owner = tracked.entry_for(path)?;
+    path.ancestors()
+        .take_while(|ancestor| ancestor.starts_with(&owner.path) && *ancestor != owner.path)
+        .find(|ancestor| ancestor.join(".git").exists())
+        .map(Path::to_path_buf)
+}
+
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/replay.rs:1323]: Correct, and fixed in eb7964810. `classify` now consults `coverage.nested` before anything else: a repository the checkpoint recorded as skipped is not something it ever held. The filesystem check finds one that still has its `.git`; the record is what still answers after the user removes it — which is the documented way to turn such a directory into ordinary content, and without this that turned the next rollback into a deletion of files the checkpoint never had. `e2e/cli/test_dotfiles_nested_repository` now removes the `.git` and rolls back to a checkpoint that skipped the directory, asserting the files survive and the run says why.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/checkpoint.rs [hunk 16]
```diff
@@ -615,7 +615,16 @@ impl Store {
...
+            // written to the tree for them, and the shared commit trailer
+            // is a format older clients parse with `deny_unknown_fields`,
+            // so adding a field for them would stop those clients reading
+            // this history at all. The list stays in this machine's own
+            // record, which is where a rollback here consults it.
+            let nested = std::mem::take(&mut checkpoint.tree.coverage.nested);
             checkpoint = repo.read_meta(&commit)?;
+            checkpoint.tree.coverage.nested = nested;
         }
         store::write_meta_cache_in(&self.state_dir, &checkpoint)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/checkpoint.rs:642]: Accurate, and it is a deliberate trade-off rather than something I can fix here — documented as such.

The list cannot go in the shared commit trailer: `CommitRecord` is parsed with `deny_unknown_fields`, so adding a field there stops released clients reading this history after a pull. That is a worse outcome than an incomplete historical record, so the list lives in this machine's checkpoint cache, and rebuilding that cache from Git does lose it.

What is lost is bounded. Live reporting is unaffected: `paths`, `status`, `save` and `track` find nested repositories by looking at the filesystem, not at the record. A rollback after a rebuild falls back to the same filesystem check, which still recognises any repository that still has its `.git` — the record only adds cover for one that has since lost it. docs/history.md says the record is best effort and what happens when it is gone.

Carrying it durably would need either a format change older clients reject or a second machine-local store keyed by checkpoint, which is a bigger design than this PR should take on.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/store.rs [hunk 17]
```diff
@@ -374,11 +374,23 @@ impl Checkpoint {
...
+            // checkpoint "did not hold" — which is how a rollback
+            // deletes them. As an omission it reads, on this mise and on
+            // an older one, as what it is: a path this commit did not
+            // capture.
             omitted: self
                 .tree
                 .coverage
                 .omitted
                 .iter()
+                .chain(self.tree.coverage.nested.iter())
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/store.rs:407]: Right — the path ended up in both lists and every report said it twice. Fixed in 3980a437d: when the local record is restored after the Git round trip, the generic copy is dropped from `omitted`, so there is one skip with the reason that can be acted on ("track it directly to capture its working files") rather than that plus "not captured in this commit".

The trailer still carries it as an omission, which is how another machine — or this one after rebuilding its index — learns about the skip at all, since the commit trailer is format-frozen for released clients and cannot gain a field. `mise dot history show` is asserted to print it exactly once, and mutation-verified: without the retain the same test reports two.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: unavailable.*

File: src/system/history/checkpoint.rs [hunk 18]
```diff
@@ -615,7 +615,16 @@ impl Store {
...
+            // The record is read back from Git so every machine reads the
+            // same one. Skipped repositories are the exception: nothing is
+            // written to the tree for them, and the shared commit trailer
+            // is a format older clients parse with `deny_unknown_fields`,
+            // so adding a field for them would stop those clients reading
+            // this history at all. The list stays in this machine's own
+            // record, which is where a rollback here consults it.
+            let nested = std::mem::take(&mut checkpoint.tree.coverage.nested);
             checkpoint = repo.read_meta(&commit)?;
+            checkpoint.tree.coverage.nested = nested;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/checkpoint.rs:640]: Right — the path ended up in both lists and every report said it twice. Fixed in 3980a437d: when the local record is restored after the Git round trip, the generic copy is dropped from `omitted`, so there is one skip with the reason that can be acted on ("track it directly to capture its working files") rather than that plus "not captured in this commit".

The trailer still carries it as an omission, which is how another machine — or this one after rebuilding its index — learns about the skip at all, since the commit trailer is format-frozen for released clients and cannot gain a field. `mise dot history show` is asserted to print it exactly once, and mutation-verified: without the retain the same test reports two.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: unavailable.*

File: src/system/history/checkpoint.rs [hunk 19]
```diff
@@ -615,7 +615,26 @@ impl Store {
...
+            // The trailer carried each of them as an ordinary omission,
+            // which is how another machine learns about the skip at all.
+            // On this one the better answer has just come back, so the
+            // generic copy goes: one skip, said once, with the reason
+            // that can be acted on.
+            checkpoint
+                .tree
+                .coverage
+                .omitted
+                .retain(|omitted| !nested.iter().any(|skip| skip.path == omitted.path));
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/checkpoint.rs]: Fixed in 6623d0886.

Verified against the code before changing anything, and the mechanism holds. `nested` comes straight from the walk, where `walk.nested` is built with `display_path` (`tracked.rs`), so it carries the platform separator. `omitted` has just been reassigned from `repo.read_meta(&commit)`, where it is rebuilt with `tree_path_to_display` (`shadow.rs`), which always writes `/`. On Windows the same skip therefore reaches the comparison as `~\.native\plugin` and `~/.native/plugin`, the exact `==` fails, and the generic omission stays — so every report names the skip twice, which is the duplicate the drop exists to remove.

`display_paths_equal` now compares display paths by component, the same way `display_under` already checks a boundary on either separator, and the drop itself is a named function, `drop_nested_from_omitted`, with a unit test that feeds it one spelling on each side. The test fails when the comparison is put back to `==`, so it guards the Windows case from Linux and macOS too.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/reconcile.rs [hunk 20]
```diff
@@ -322,9 +358,31 @@ pub(crate) fn reconcile(
...
+                .is_some_and(|object| is_gitlink(object.as_ref()))
+        {
+            let version = plan.apply.take().flatten();
+            plan.skipped = Some(NESTED_NOT_SHARED.into());
+            plan.next.acknowledged = version.clone();
+            plan.next.reconciled = version.clone();
+            plan.next.applied = version;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/reconcile.rs:387]: Traced this against the code and it holds, for the pointer-versus-absent branch at least. Leaving it for @jdx rather than changing it here, because both ways out trade against the property that branch was written to keep.

The case: machine A has a nested repository at `~/.sample/plugin` and publishes the pointer; machine B has nothing there. On B, `s = None` and `t = gitlink`, so `pointer_or_absent` holds on both sides and B records `acknowledged = reconciled = applied = t_version` — the pointer — although B has nothing on disk and wrote nothing. Later A replaces that path with ordinary files. On B, `local_changed = s_version != a` is now `None != Some(pointer)` = true, `remote_changed` is true, so it takes the `(true, true)` arm and calls `merge(base = pointer, ours = None, theirs = files)`, which is `(None, Some(_)) => ModifyDelete`. B gets a conflict describing a local deletion it never made, and cannot receive the files without resolving it. With `a = None` the same sync takes `(false, true)` and applies the files cleanly.

The other half of the report is less clear to me. Where `s` is itself the pointer, `local_changed` is false and the `(false, true)` arm applies the files with no merge and no lookup of the pointer object, so I could not reproduce an error there.

What I am not sure of is the remedy, which is why I am not writing one. Recording `s_version` instead of `t_version` in that branch fixes the false conflict and still leaves the plan a no-op — `plan.next == record`, so nothing is pushed and nothing waits or blocks publication. But `skip_pointer_applications` records the same pointer for a different reason: an application that would write a gitlink, including one a conflict resolution chose, and there "clearing `apply`" alone leaves `plan.next != record`, so the plan persists rather than settling. Whether the answer is to acknowledge what this machine actually has, or to acknowledge the decision without using the pointer object as the merge base, is a call about what the recorded state is supposed to mean, and that belongs with the maintainer rather than with a review-comment fix.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/tracked.rs [hunk 21]
```diff
@@ -633,6 +653,19 @@ pub(crate) fn is_builtin_credential(path: &Path, name: &str) -> bool {
...
+/// **The same path reaches this comparison in two spellings.** A walk
+/// writes display paths with [`crate::file::display_path`], which uses
+/// the platform separator, and a record read back out of a commit
+/// rebuilds them with [`tree_path_to_display`], which always writes `/`.
+/// On Windows one path is then `~\.codex\plugin` and the other
+/// `~/.codex/plugin`, and comparing the strings says they are two
+/// different paths.
+pub(crate) fn display_paths_equal(left: &str, right: &str) -> bool {
+    left.split(['/', '\\']).eq(right.split(['/', '\\']))
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/tracked.rs]: You are right, and the separator fold is gone. Fixed in 0e82facd6.

I had folded `/` and `\` at the comparison, which does not make the two spellings meet: `display_user` only rewrites `$HOME` to `~` under `cfg!(unix)`, so on Windows the walk writes `C:\Users\me\.native\plugin` while the rebuilt record holds `~/.native/plugin`, and the home prefix differs whatever is done about separators. Greptile's point on the same change was the other half — on unix a backslash is an ordinary character in a filename, so folding called `~/a/b` and a file named `a\b` one path.

Neither could be settled at the comparison, so the writers settle it. `Checkpoint::portable_path` is now a method rather than a closure inside `for_commit`, and the skip is matched on `tree_path_to_display(portable_path(..))` — the same two conversions the record itself went through — with `==`. Nothing is folded and nothing is normalized at the comparison.

I did not merge `display_path` and `tree_path_to_display` into one spelling, which was the other way out: `display_user` feeds strings that are matched and not only shown — `ToolRequest::Path::version()` builds `path:<display_user>` and lockfile entries are matched on it — so changing its home handling would change identity rather than presentation.

The test builds its coverage the way a capture does, asserts the drop key equals what the trailer round trip produces, and on Windows additionally asserts that the walked spelling is *not* that key, so the derivation is exercised rather than assumed.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/tracked.rs [hunk 22]
```diff
@@ -633,6 +653,19 @@ pub(crate) fn is_builtin_credential(path: &Path, name: &str) -> bool {
...
+///
+/// **The same path reaches this comparison in two spellings.** A walk
+/// writes display paths with [`crate::file::display_path`], which uses
+/// the platform separator, and a record read back out of a commit
+/// rebuilds them with [`tree_path_to_display`], which always writes `/`.
+/// On Windows one path is then `~\.codex\plugin` and the other
+/// `~/.codex/plugin`, and comparing the strings says they are two
+/// different paths.
+pub(crate) fn display_paths_equal(left: &str, right: &str) -> bool {
+    left.split(['/', '\\']).eq(right.split(['/', '\\']))
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/tracked.rs]: Correct, and the fold is gone. Fixed in 0e82facd6.

On unix a backslash is an ordinary character in a filename, so folding it with `/` made `~/.native/plugin` and a file genuinely named `.native\plugin` compare equal. Bugbot flagged the opposite end of the same change: folding did not fix Windows either, because `display_user` only shortens `$HOME` to `~` under `cfg!(unix)`, so the two spellings still differed in their prefix.

A comparison cannot satisfy both from the strings alone, so the two writers now produce the same string instead. The skip is matched on `tree_path_to_display(Checkpoint::portable_path(..))`, the same pair of conversions the rebuilt record went through, compared with `==`. The test includes the case you raise: a path holding a literal backslash is not dropped by a rule that names the slash-separated path of the same name.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5; version: 2.1.270.*

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
